src/fromager/bootstrapper.py

```diff
@@ -194,105 +210,33 @@ def bootstrap(self, req: Requirement, req_type: RequirementType) -> Version:
-                    f"'{sdist_root_dir}/../..' should be {self.ctx.work_dir}"
-                )
-            unpack_dir = sdist_root_dir.parent
-
-            build_env = build_environment.BuildEnvironment(
-                ctx=self.ctx,
-                parent_dir=sdist_root_dir.parent,
+            # Look for an existing wheel in caches (3 levels: build, downloads,
+            # cache server) before building from source.
+            cached_wheel_filename, unpacked_cached_wheel = self._find_cached_wheel(
+                req, resolved_version
             )
 
-            # need to call this function irrespective of whether we had the wheel cached
-            # so that the build dependencies can be bootstrapped
-            self._prepare_build_dependencies(req, sdist_root_dir, build_env)
+            # Build from source (download, prepare, build wheel/sdist)
+            build_result = self._build_from_source(
+                req=req,
+                resolved_version=resolved_version,
+                source_url=source_url,
+                build_sdist_only=build_sdist_only,
+                cached_wheel_filename=cached_wheel_filename,
+                unpacked_cached_wheel=unpacked_cached_wheel,
+            )
 
-            if cached_wheel_filename:
-                logger.debug(
-                    f"getting install requirements from cached "
-                    f"wheel {cached_wheel_filename.name}"
-                )
-                # prefer existing wheel even in sdist_only mode
-                # skip building even if it is a non-fromager built wheel
-                wheel_filename = cached_wheel_filename
-                build_sdist_only = False
-            elif build_sdist_only:
-                # get install dependencies from sdist and pyproject_hooks (only top-level and install)
-                logger.debug(
-                    f"getting install requirements from sdist "
-                    f"{req.name}=={resolved_version} ({req_type})"
-                )
-                wheel_filename = None
-                sdist_filename = self._build_sdist(
-                    req, resolved_version, sdist_root_dir, build_env
-                )
-            else:
-                # build wheel (build requirements, full build mode)
-                logger.debug(
-                    f"building wheel {req.name}=={resolved_version} "
-                    f"to get install requirements ({req_type})"
-                )
-                wheel_filename, sdist_filename = self._build_wheel(
-                    req, resolved_version, sdist_root_dir, build_env
-                )
+            # Unpack result
+            wheel_filename = build_result.wheel_filename
+            sdist_filename = build_result.sdist_filename
+            unpack_dir = build_result.unpack_dir
+            sdist_root_dir = build_result.sdist_root_dir
+            build_env = build_result.build_env
+            source_url_type = build_result.source_url_type
```

Comment:
Is it really necessary to unpack the data class? How about you use `build_result.sdist_filename` and other attributes throughout the rest of the function?